src/republication.rs

```diff
@@ -0,0 +1,164 @@
+};
+use heapless::Vec;
+
+use embedded_nal::TcpClientStack;
+
+pub(crate) struct RepublicationBuffer<'a> {
+    publish_buffer: RingBuffer<'a>,
+    pending_pubrel: Vec<(u16, ReasonCode), 10>,
+    republish_index: Option<usize>,
+    pubrel_republish_index: Option<usize>,
+    max_tx_size: usize,
+}
+
+impl<'a> RepublicationBuffer<'a> {
+    pub fn new(buf: &'a mut [u8], max_tx_size: usize) -> Self {
+        Self {
+            publish_buffer: RingBuffer::new(buf),
+            pending_pubrel: Vec::new(),
+            republish_index: None,
+            pubrel_republish_index: None,
+            max_tx_size,
+        }
+    }
+
+    pub fn clear(&mut self) {
+        self.republish_index.take();
+        self.pubrel_republish_index.take();
+        self.publish_buffer.clear();
+        self.pending_pubrel.clear();
+    }
+
+    pub fn reset(&mut self) {
+        self.republish_index.replace(0);
+        self.pubrel_republish_index.replace(0);
+
+        // If there's nothing to republish, clear out our states.
+        if self.pending_pubrel.is_empty() {
+            self.pubrel_republish_index.take();
+        }
+
+        if self.publish_buffer.len() == 0 {
+            self.republish_index.take();
+        }
+    }
+
+    pub fn pop_publish(&mut self, id: u16) -> Result<(), ProtocolError> {
+        let (_, header) = self
+            .publish_buffer
+            .probe_header(0)
+            .ok_or(ProtocolError::BadIdentifier)?;
+        if header.packet_id != id {
+            return Err(ProtocolError::BadIdentifier);
+        }
+
+        if let Some(index) = self.republish_index.take() {
+            if index > 1 {
+                self.republish_index.replace(index - 1);
+            }
+        }
+
+        self.publish_buffer.pop(header.len);
+        Ok(())
+    }
+
+    pub fn push_publish(&mut self, packet: &[u8]) -> Result<(), ProtocolError> {
+        if self.publish_buffer.push_slice(packet).is_some() {
+            return Err(ProtocolError::BufferSize);
+        }
+
+        Ok(())
+    }
+
+    pub fn pop_pubrel(&mut self, id: u16) -> Result<(), ProtocolError> {
+        // We always have to pop from the front of the vector to enforce FIFO characteristics.
+        let Some((pending_id, _)) = self.pending_pubrel.get(0) else {
+            return Err(ProtocolError::UnexpectedPacket);
+        };
+
+        if *pending_id != id {
+            return Err(ProtocolError::BadIdentifier);
+        }
+
+        // Now that we received the PubComp for this PubRel, we can remove it from our session
+        // state. We will not need to retransmit this upon reconnection.
+        self.pending_pubrel.remove(0);
+
+        if let Some(index) = self.pubrel_republish_index.take() {
+            if index > 1 {
+                self.pubrel_republish_index.replace(index - 1);
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn push_pubrel(&mut self, pubrel: &PubRel) -> Result<(), ProtocolError> {
+        self.pending_pubrel
```

Comment:
The `pubrel` list is now a fixed size, but `publish` is variable in length. This means that the `pubrel` list could fill up if we send out too many `publish` messages to the broker. Perhaps we should encode the pending `pubrel` messages directly into the ring buffer as well for republication..

However, I don't think MQTT guarantees ordering between the PUBLISH and the PUBREL messages, so we might get a PUBCOMP message before we get a PUBREC if we send a PUBLISH followed by a PUBREL to the server. Because the ring buffer requires ordering, this would break things.... Need to think about this further

Reply:
That's a really good point, thanks for pointing that out. Simplifies the design.